### PR TITLE
Add new xmldata BBE to the conversion between XML and record

### PR DESCRIPTION
**Pull request by kalaiyarasiganeshalingam**
## Purpose
$Subject

Fixes: https://github.com/ballerina-platform/ballerina-standard-library/issues/3114

### CODE REVIEW

File: examples/xml-record-conversion/xml_record_conversion.bal [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    int count;
+};
+
+@xmldata:Namespace {
+    uri: "http://example.com"
+}
+type ItemCode record {
+    @xmldata:Attribute
+    string discount;
+    string \#content;
```
Comment (daneshk) [inline examples/xml-record-conversion/xml_record_conversion.bal:31]: Don't put `\#content` here. This is a special case. we don't need to cover everything in one BBE

File: examples/xml-record-conversion/xml_record_conversion.bal [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+};
+
+public function main() returns error? {
+    // Converts a `record` representation to its XML representation.
+    xml result = check xmldata:toXml(data);
+    io:println(result);
+
+    // Converts an XML representation to its `record` representation.
+    Invoice output = check xmldata:fromXml(result);
+    io:println(output);
```
Comment (daneshk) [inline examples/xml-record-conversion/xml_record_conversion.bal:57]: This should be in two different BBEs. one for record -> XML, other for XML -> record

File: examples/xml-record-conversion/xml_record_conversion.bal [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            itemCode: {discount: "22%", \#content: "200777"},
+            count: 7
+        }
+    ],
+    address: "20, Palm grove, Colombo 3",
+    attr: "attr-val",
+    'xmlns: "http://sample.com",
+    ns\:attr: "http://attr.com"
+};
```
Comment (daneshk) [inline examples/xml-record-conversion/xml_record_conversion.bal:48]: this should go inside the main function. record definition can put outside

File: examples/xml-record-conversion/xml_record_conversion.bal [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    @xmldata:Attribute
+    string 'xmlns;
+    @xmldata:Attribute
+    string attr?;
+    @xmldata:Attribute
+    string ns\:attr?;
+};
+
+type Item record {
+    string|ItemCode itemCode;
```
Comment (daneshk) [inline examples/xml-record-conversion/xml_record_conversion.bal:21]: why do we need union type here

File: examples/xml-record-conversion/xml_record_conversion.bal (hunk 2 above)
Comment (kalaiyarasiganeshalingam) [reply examples/xml-record-conversion/xml_record_conversion.bal:57]: We are maintaining a BBE for the conversion between  XML and JSON. So I also added one for this. 
https://github.com/ballerina-platform/ballerina-distribution/tree/master/examples/xml-json-conversion

Comment (daneshk) [reply examples/xml-record-conversion/xml_record_conversion.bal:57]: We need to rethink them later. We need to make them as simple as possible

File: examples/xml-record-conversion/xml_record_conversion.bal (hunk 4 above)
Comment (kalaiyarasiganeshalingam) [reply examples/xml-record-conversion/xml_record_conversion.bal]: Removed and simplified this.

File: examples/xml-from-record-conversion/xml_from_record_conversion.bal [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+@xmldata:Name {
+    value: "purchased_item"
+}
+type Item record {
+    string itemCode;
+    @xmldata:Name {
+        value: "item_count"
+    }
+    int count;
+};
```
Comment (daneshk) [inline examples/xml-from-record-conversion/xml_from_record_conversion.bal:30]: Shall we remove Name annotation from the example

Comment (kalaiyarasiganeshalingam) [reply examples/xml-from-record-conversion/xml_from_record_conversion.bal:24]: Removed [6999224](https://github.com/ballerina-platform/ballerina-distribution/pull/2963/commits/69992249f0ae1503bb90d6bcaa0f083fba511e7e)

File: examples/xml-from-record-conversion/xml_from_record_conversion.bal [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+// Defines a record type with annotations.
+@xmldata:Namespace {
+    prefix: "ns",
+    uri: "http://sdf.com"
+}
+type Invoice record {
+    int id;
+    Item[] items;
+    @xmldata:Attribute
+    string 'xmlns?;
```
Comment (daneshk) [inline examples/xml-from-record-conversion/xml_from_record_conversion.bal:13]: Normally this xmlns value is knowns when defining the record right? it is not changed in value. Shall we add the value here

File: examples/xml-from-record-conversion/xml_from_record_conversion.bal [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+public function main() returns error? {
+    // Creates an `Invoice` records.
+    Invoice data = {
+        id: 1,
+        items: [
+            {itemCode: "223345", count: 1},
+            {itemCode: "223300", count: 7}
+        ],
+        attr: "attr-val"
```
Comment (daneshk) [inline examples/xml-from-record-conversion/xml_from_record_conversion.bal:34]: shall we have a meaningful attribute value here?

File: examples/xml-from-record-conversion/xml_from_record_conversion.bal [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    Item[] items;
+    @xmldata:Attribute
+    string 'xmlns = "example.com";
+    @xmldata:Attribute
+    string attr?;
+};
+
+@xmldata:Namespace {
+    uri: "http://example1.com"
+}
```
Comment (daneshk) [inline examples/xml-from-record-conversion/xml_from_record_conversion.bal:20]: Shall we not add namespace annotation for both record types?

File: examples/xml-from-record-conversion/xml_from_record_conversion.bal [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+type Item record {
+    string itemCode;
+    int count;
+};
+
+public function main() returns error? {
+    // Creates an `Invoice` records.
```
Comment (praneesha) [inline examples/xml-from-record-conversion/xml_from_record_conversion.bal:27]:
```suggestion
    // Creates an `Invoice` record.
```

File: examples/xml-from-record-conversion/xml_from_record_conversion.description [hunk 10]
```diff
@@ -0,0 +1,3 @@
+// The `xmldata` library provides APIs to perform conversion from Ballerina record/map<anydata> to XML.
```
Comment (praneesha) [inline examples/xml-from-record-conversion/xml_from_record_conversion.description:1]:
```suggestion
// The `xmldata` library provides APIs to perform conversion from Ballerina `record/map<anydata>` to XML.
```

File: examples/xml-from-record-conversion/xml_from_record_conversion.description [hunk 11]
```diff
@@ -0,0 +1,3 @@
...
+// For more information on the underlying module,
+// see the [xmldata module](https://docs.central.ballerina.io/ballerina/xmldata/latest/).
```
Comment (praneesha) [inline examples/xml-from-record-conversion/xml_from_record_conversion.description:3]:
```suggestion
// see the [`xmldata` module](https://docs.central.ballerina.io/ballerina/xmldata/latest/).
```

File: examples/xml-from-record-conversion/xml_from_record_conversion.md [hunk 12]
```diff
@@ -0,0 +1,11 @@
+# XML from record conversion
+
+The `xmldata` library provides APIs to perform conversion from Ballerina record/map<anydata> to XML.
```
Comment (praneesha) [inline examples/xml-from-record-conversion/xml_from_record_conversion.md:3]:
```suggestion
The `xmldata` library provides APIs to perform conversion from `Ballerina record/map<anydata>` to XML.
```

File: examples/xml-from-record-conversion/xml_from_record_conversion.md [hunk 13]
```diff
@@ -0,0 +1,11 @@
+# XML from record conversion
```
Comment (praneesha) [inline examples/xml-from-record-conversion/xml_from_record_conversion.md:1]:
```suggestion
# XML from record conversion
```
This title is confusing. Should describe from what to what is the conversion.

File: examples/xml-from-record-conversion/xml_from_record_conversion.md [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+
+The `xmldata` library provides APIs to perform conversion from Ballerina record/map<anydata> to XML.
+
+For more information on the underlying module, see the [xmldata module](https://docs.central.ballerina.io/ballerina/xmldata/latest/).
```
Comment (praneesha) [inline examples/xml-from-record-conversion/xml_from_record_conversion.md:5]:
```suggestion
For more information on the underlying module, see the [`xmldata` module](https://docs.central.ballerina.io/ballerina/xmldata/latest/).
```

File: examples/xml-to-record-conversion/xml_to_record_conversion.description [hunk 15]
```diff
@@ -0,0 +1,3 @@
+The `xmldata` library provides API to perform conversion from XML to Ballerina record/map<anydata>.
```
Comment (praneesha) [inline examples/xml-to-record-conversion/xml_to_record_conversion.description:1]:
```suggestion
The `xmldata` library provides an API to perform the conversion from XML to Ballerina `record/map<anydata>`.
```

File: examples/xml-to-record-conversion/xml_to_record_conversion.description [hunk 16]
```diff
@@ -0,0 +1,3 @@
...
+// For more information on the underlying module,
+// see the [xmldata module](https://docs.central.ballerina.io/ballerina/xmldata/latest/).
```
Comment (praneesha) [inline examples/xml-to-record-conversion/xml_to_record_conversion.description:3]:
```suggestion
// see the [`xmldata` module](https://docs.central.ballerina.io/ballerina/xmldata/latest/).
```

File: examples/xml-to-record-conversion/xml_to_record_conversion.md [hunk 17]
```diff
@@ -0,0 +1,11 @@
+# XML to record conversion
+
+The `xmldata` library provides API to perform conversion from XML to Ballerina record/map<anydata>.
```
Comment (praneesha) [inline examples/xml-to-record-conversion/xml_to_record_conversion.md:3]:
```suggestion
The `xmldata` library provides an API to perform conversion from XML to Ballerina `record/map<anydata>`.
```

File: examples/xml-to-record-conversion/xml_to_record_conversion.md [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
+
+For more information on the underlying module, see the [xmldata module](https://docs.central.ballerina.io/ballerina/xmldata/latest/).
```
Comment (praneesha) [inline examples/xml-to-record-conversion/xml_to_record_conversion.md:5]:
```suggestion
For more information on the underlying module, see the [`xmldata` module](https://docs.central.ballerina.io/ballerina/xmldata/latest/).
```

File: gradle.properties [hunk 19]
```diff
@@ -3,7 +3,7 @@ group=org.ballerinalang
 # During the release workflow, the following will get bumped automatically
```
Comment (praneesha) [inline gradle.properties:3]:
```suggestion
# During the release workflow, the following will get bumped automatically.
```

File: examples/xml-from-record-conversion/xml_from_record_conversion.md (hunk 13 above)
Comment (kalaiyarasiganeshalingam) [reply examples/xml-from-record-conversion/xml_from_record_conversion.md]: I have changed the title to `Conversion of Ballerina record to XML`.

File: examples/xml-from-record-conversion/xml_from_record_conversion.bal (hunk 8 above)
Comment (kalaiyarasiganeshalingam) [reply examples/xml-from-record-conversion/xml_from_record_conversion.bal:20]: No need to add a namespace annotation for both record types. If we want to add `xmlns` with a different URL in the child sequence, we can add it in the child type like this.

File: examples/xml-from-record-conversion/xml_from_record_conversion.md (hunk 13 above)
Comment (daneshk) [reply examples/xml-from-record-conversion/xml_from_record_conversion.md]: This name should consistent with the XML-JSON conversion

Comment (daneshk) [reply examples/xml-from-record-conversion/xml_from_record_conversion.md]: Let's name the BBEs like,

XML to JSON conversion
JSON to XML conversion
XML to Record conversion
Record to XML conversion

**Review by praneesha — APPROVED** [review]

**Pull request merged**